### PR TITLE
Fix heap-buffer-overflow in ossl_i2c_ASN1_BIT_STRING, when data contains only zero values

### PR DESCRIPTION
**Pull request by AITsygunka**
Fix heap-buffer-overflow in ```ossl_i2c_ASN1_BIT_STRING()``` in ```crypto/asn1/a_bitstr.c```
Issue  #26189

CLA: trivial

### CODE REVIEW

**Comment by t8m** [comment]
Please drop the merge commits (`git reset --hard HEAD~~; git push --force`)

**Review by t8m — APPROVED** [review]

**Comment by davidben** [comment]
I don't think this fix is quite correct. What's going on here is that OpenSSL's `ASN1_STRING` representation for BIT STRINGs is pretty bizarre and has two modes:

1. `ASN1_STRING_FLAG_BITS_LEFT` set, where the number of bits is exactly what is represented.
2. `ASN1_STRING_FLAG_BITS_LEFT` unset, where trailing zero bits are implicitly chopped off

This code is evaluating the actual length of the second one and the bug is that an all zero `ASN1_STRING` underflows. But rather than an error, it should probably correctly chop off all bits and encode as an empty BIT STRING.

TBH, mode 2 seems just generally a bad idea and probably should be removed. There are a lot of weird things about it, like the fact that equality doesn't work right. We haven't done it yet, but I've been considering whether moving the implicit chopping to `ASN1_BIT_STRING_set_bit` is compatible enough in practice. The problem is OpenSSL's BIT STRING abstraction is incredibly leaky and actually setting a BIT STRING is remarkably fiddly because you need to mess with `flags`.

**Comment by davidben** [comment]
That is, I think the correct fix would be to skip reading `a->data[len - 1]` and then proceed with `bits = 0`.

Also, it seems worth adding a test for this.

File: crypto/asn1/a_bitstr.c [hunk 1]
```diff
@@ -38,27 +38,28 @@ int ossl_i2c_ASN1_BIT_STRING(ASN1_BIT_STRING *a, unsigned char **pp)
...
-                bits = 4;
-            else if (j & 0x20)
-                bits = 5;
-            else if (j & 0x40)
-                bits = 6;
-            else if (j & 0x80)
-                bits = 7;
-            else
-                bits = 0;       /* should not happen */
+            else {
```
Comment (t8m) [inline crypto/asn1/a_bitstr.c:42]: Please add `{}` around the if block.

**Comment by AITsygunka** [comment]
> @AITsygunka could you do what @davidben suggested above? Unfortunately adding a test would not be acceptable with CLA: trivial so you would have to either submit a CLA or just update the fix and we can add the test later.

@t8m i updated the previous my fix as recommended by @davidben. But I'm not sure I did it right. Please review it.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
OK with CLA: trivial

**Review by tom-cosgrove-arm — APPROVED** [review]
Okay with trivial, although a test added later would be good.

Would be good to squash commits when this is merged

**Comment by t8m** [comment]
Squashed and merged to all the active branches. Thank you for your contribution.

**Pull request closed without merging**